File: src/finance/finance.controller.ts
import { Controller, Get, Post, Put, Delete, Body, Param, UseGuards } from '@nestjs/common';
import { ApiTags, ApiBearerAuth } from '@nestjs/swagger';
import { FinanceService } from './finance.service';
import { JwtAuthGuard } from '../auth/guards/jwt.guard';
import { BandGuard } from '../common/guards/band.guard';
import { CurrentUser } from '../common/decorators/current-user.decorator';

@ApiTags('finance')
@ApiBearerAuth('JWT')
@Controller('finance')
@UseGuards(JwtAuthGuard, BandGuard)
export class FinanceController {
  constructor(private readonly finance: FinanceService) {}

  // -- Transactions --

  @Get('transactions')
  findAllTransactions(@CurrentUser() user: any) {
    return this.finance.findAllTransactions(user.bandId);
  }

  @Post('transactions')
  createTransaction(@CurrentUser() user: any, @Body() dto: any) {
    return this.finance.createTransaction(user.bandId, dto);
  }

  @Put('transactions/:id')
  updateTransaction(@CurrentUser() user: any, @Param('id') id: string, @Body() dto: any) {
    return this.finance.updateTransaction(user.bandId, id, dto);
  }

  @Delete('transactions/:id')
  removeTransaction(@CurrentUser() user: any, @Param('id') id: string) {
    return this.finance.removeTransaction(user.bandId, id);
  }

  // -- Wish List --

  @Get('wish-list')
  findAllWishList(@CurrentUser() user: any) {
    return this.finance.findAllWishList(user.bandId);
  }

  @Post('wish-list')
  createWishListItem(@CurrentUser() user: any, @Body() dto: any) {
    return this.finance.createWishListItem(user.bandId, dto);
  }

  @Put('wish-list/:id')
  updateWishListItem(@CurrentUser() user: any, @Param('id') id: string, @Body() dto: any) {
    return this.finance.updateWishListItem(user.bandId, id, dto);
  }

  @Delete('wish-list/:id')
  removeWishListItem(@CurrentUser() user: any, @Param('id') id: string) {
    return this.finance.removeWishListItem(user.bandId, id);
  }

  // -- Initial Balance --

  @Get('balance')
  getInitialBalance(@CurrentUser() user: any) {
    return this.finance.getInitialBalance(user.bandId);
  }

  @Put('balance')
  setInitialBalance(@CurrentUser() user: any, @Body() body: { initialBalance: number }) {
    return this.finance.setInitialBalance(user.bandId, body.initialBalance);
  }
}

File: spring-src/main/java/com/blackout/api/finance/infrastructure/web/FinanceController.java
package com.blackout.api.finance.infrastructure.web;

import com.blackout.api.finance.application.service.FinanceApplicationService;
import com.blackout.api.finance.infrastructure.web.dto.*;
import com.blackout.api.shared.infrastructure.security.BlackoutAuthentication;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/finance")
public class FinanceController {

    private final FinanceApplicationService service;

    public FinanceController(FinanceApplicationService service) {
        this.service = service;
    }

    // ── Transactions ──────────────────────────────────────────────────────────

    @GetMapping("/transactions")
    public List<TransactionResponse> getTransactions(BlackoutAuthentication auth) {
        return service.findAllTransactions(auth.getBandId());
    }

    @PostMapping("/transactions")
    @ResponseStatus(HttpStatus.CREATED)
    public TransactionResponse createTransaction(
            BlackoutAuthentication auth,
            @Valid @RequestBody CreateTransactionRequest dto) {
        return service.createTransaction(auth.getBandId(), dto);
    }

    @PutMapping("/transactions/{id}")
    public TransactionResponse updateTransaction(
            BlackoutAuthentication auth,
            @PathVariable String id,
            @RequestBody UpdateTransactionRequest dto) {
        return service.updateTransaction(auth.getBandId(), id, dto);
    }

    @DeleteMapping("/transactions/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteTransaction(BlackoutAuthentication auth, @PathVariable String id) {
        service.removeTransaction(auth.getBandId(), id);
    }

    // ── Wish List ─────────────────────────────────────────────────────────────

    @GetMapping("/wish-list")
    public List<WishListItemResponse> getWishList(BlackoutAuthentication auth) {
        return service.findAllWishList(auth.getBandId());
    }

    @PostMapping("/wish-list")
    @ResponseStatus(HttpStatus.CREATED)
    public WishListItemResponse createWishListItem(
            BlackoutAuthentication auth,
            @Valid @RequestBody CreateWishListItemRequest dto) {
        return service.createWishListItem(auth.getBandId(), dto);
    }

    @PutMapping("/wish-list/{id}")
    public WishListItemResponse updateWishListItem(
            BlackoutAuthentication auth,
            @PathVariable String id,
            @RequestBody UpdateWishListItemRequest dto) {
        return service.updateWishListItem(auth.getBandId(), id, dto);
    }

    @DeleteMapping("/wish-list/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteWishListItem(BlackoutAuthentication auth, @PathVariable String id) {
        service.removeWishListItem(auth.getBandId(), id);
    }

    // ── Initial Balance ───────────────────────────────────────────────────────

    @GetMapping("/initial-balance")
    public InitialBalanceResponse getInitialBalance(BlackoutAuthentication auth) {
        return service.getInitialBalance(auth.getBandId());
    }

    @PutMapping("/initial-balance")
    public InitialBalanceResponse setInitialBalance(
            BlackoutAuthentication auth,
            @RequestBody InitialBalanceRequest dto) {
        return service.setInitialBalance(auth.getBandId(), dto.initialBalance());
    }
}
